locator/abstract_replication_strategy.cc

```diff
@@ -330,14 +330,67 @@ abstract_replication_strategy::get_pending_address_ranges(const token_metadata_p
 
 future<mutable_vnode_effective_replication_map_ptr> calculate_effective_replication_map(replication_strategy_ptr rs, token_metadata_ptr tmptr) {
     replication_map replication_map;
+    ring_mapping pending_endpoints;
+    ring_mapping read_endpoints;
     const auto& sorted_tokens = tmptr->sorted_tokens();
 
     if (!sorted_tokens.empty()) {
         replication_map.reserve(sorted_tokens.size());
         if (rs->natural_endpoints_depend_on_token()) {
```

Comment:
> In this commit we add logic to calculate pending_endpoints and
read_endpoints, similar to how it was done in update_pending_ranges.
If natural_endpoints_depend_on_token is false, we leave them empty,
since the set of nodes does not depend on the token.

`natural_endpoints_depend_on_token` only means that for each vnode/token the set of natural endpoints will be the same. But it still depends on the tokens in the token ring. I.e. in `calculate_natural_endpoints`(t, *tmptr)` it still depends on `*tmptr` but not on `t`.

With pending endpoints and read endpoints it's the same. We still depend on the token ring. We cannot leave them empty!

Let's take `EverywhereReplicationStrategy`. The token ring is `10 ->A`. Node `B` is bootstrapping with token `20` so the token ring will eventually be `10->A, 20->B`.

Natural endpoints (for any token) will be {A}. Pending endpoints will be {B}, not {}. In `write_both_read_new`, read endpoints will be {A, B}, not {} (or {A}).

Please add a test for `EverywhereReplicationStrategy`.

Reply:
Thanks. Another stupid mistake. You have caught a lot in these series, I should reconsider my decision making.